File: javascript/allergies/allergies.js
const N2ITEM = [
  'eggs',
  'peanuts',
  'shellfish',
  'strawberries',
  'tomatoes',
  'chocolate',
  'pollen',
  'cats',
].reduce((m, item, idx) => {
  m[Math.pow(2, idx)] = item
  return m
}, {})

const decomposeScore = score => {
  let log
  const ns = []
  while (score > 0) {
    log = Math.floor(Math.log(score) / Math.log(2))
    const n = Math.pow(2, log)
    ns.push(n)
    score -= n
  }
  return ns.reverse()
}

const takeWhile = (arr, p) => {
  const ret = []
  for (let i = 0; p(arr[i]) && i < arr.length; i++) {
    ret.push(arr[i])
  }
  return ret
}

class Allergies {
  constructor(score, n2item = N2ITEM) {
    const max = Number(Math.max(...Object.keys(n2item)))
    const ns = takeWhile(decomposeScore(score), n => n <= max)
    this.allergies = ns.map(n => n2item[n])
  }

  list() {
    return this.allergies
  }

  allergicTo(item) {
    return this.allergies.indexOf(item) >= 0
  }
}

module.exports = Allergies

File: go/allergies/allergies.go
package allergies

// var table = map[int]string{
// 	1: "eggs",
// }

var allergyNames = []string{
	"eggs",
	"peanuts",
	"shellfish",
	"strawberries",
	"tomatoes",
	"chocolate",
	"pollen",
	"cats",
}

func Allergies(allergies uint) []string {
	bits := toBits(allergies)
	names := make([]string, 0, len(bits))
	for i, bit := range bits {
		if bit {
			names = append(names, allergyNames[i])
		}
	}
	return names
}

func AllergicTo(allergies uint, allergen string) bool {
	bits := toBits(allergies)
	for i, bit := range bits {
		if bit && allergyNames[i] == allergen {
			return true
		}
	}
	return false
}

func toBits(n uint) []bool {
	maxBit := 8 // 128
	bits := make([]bool, 0, maxBit)
	for i := 0; n > 0 && i < maxBit; i++ {
		bit := n&1 == 1
		bits = append(bits, bit)
		n >>= 1
	}
	return bits
}
